### PR TITLE
coll/base: Prevent hang in alltoallv

### PR DESCRIPTION
**Pull request by lrbison**
This fix addresses two problems:
 - ~~Not all ranks may be executing an in-place algorithm~~ that is an invalid MPI program
 - Consistently skip 0-byte messages in all algorithms

### CODE REVIEW

**Comment by lrbison** [comment]
I tested this PR with the following test case.  Previously alltoallv hung on algorithm 2 (pairwise).

`mpirun -mca coll_tuned_use_dynamic_rules 1 -mca coll_tuned_alltoallv_algorithm {0,1,2} -n 64 ./a.out`

```
#include <mpi.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

int main(int argc, char** argv) {
    int rc = 0;

    // Initialize the MPI environment
    MPI_Init(NULL, NULL);

    int world_size;
    int world_rank;
    const int nrank_test = 5;
    MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
    MPI_Comm_size(MPI_COMM_WORLD, &world_size);

    MPI_Comm_set_errhandler(MPI_COMM_WORLD, MPI_ERRORS_RETURN );

    char *sendbuf0, *recvbuf, *sendbuf;
    int *sendcounts, *recvcounts;
    int *sdispls, *rdispls;
    sendbuf0 = calloc(world_size, sizeof(char));
    recvbuf = calloc(world_size, sizeof(char));
    sendcounts = calloc(world_size, sizeof(int));
    recvcounts = calloc(world_size, sizeof(int));
    sdispls = calloc(world_size, sizeof(int));
    rdispls = calloc(world_size, sizeof(int));
    sendbuf = sendbuf0;


    for (int i=0;i<world_size;i++) {
        sdispls[i] = i;
        rdispls[i] = i;
    }
    MPI_Datatype sendtype, recvtype;

    sendtype = MPI_CHAR;
    recvtype = MPI_CHAR;

    if (world_rank < 2) {
        /* ignored: 0 counts */
        if (world_rank%2) { sendbuf = MPI_IN_PLACE; }
    } else if (world_rank < 4) {
        /* ignored: 0 sizes */
        if (world_rank%2) { sendbuf = MPI_IN_PLACE; }
        for (int i=0; i<world_size; i++) {
            sendcounts[i] = 1; recvcounts[i] = 1;
        }
        MPI_Type_contiguous(0, MPI_CHAR, &sendtype);
        MPI_Type_contiguous(0, MPI_CHAR, &recvtype);
        MPI_Type_commit(&sendtype);
        MPI_Type_commit(&recvtype);
    } else {
        /* normal */
        for (int j=4; j<world_size; j++) {
            sendcounts[j] = 1; recvcounts[j] = 1;
        }
        if (world_rank%3==0) { sendbuf = MPI_IN_PLACE; }
    }
    // sendbuf = MPI_IN_PLACE;

    double precision = MPI_Wtick();
    double t0 = MPI_Wtime();
    char init_val = (sendbuf == MPI_IN_PLACE) ? world_rank : world_rank + 100;
    memset(recvbuf, init_val, world_size*sizeof(char));
    memset(sendbuf0, world_rank, world_size*sizeof(char));

    rc = MPI_Alltoallv(sendbuf, sendcounts,
                sdispls, sendtype, recvbuf,
                recvcounts, rdispls, recvtype,
                MPI_COMM_WORLD);
    double t1 = MPI_Wtime();

    int error_counts=0;
    for (int i=0; i<world_size; i++) {
        char expected = i;
        if (world_rank < 4 || i < 4) expected = init_val;
        int valid_data = recvbuf[i] == expected;
        if (!valid_data) {
            printf("Rank %d sees %d in position %d.  Expected %d\n", world_rank, recvbuf[i], i, i);
            error_counts++;
        }
    }

    if (rc != 0) {
        printf("Error: Rank %d completed with status=%d\n",world_rank, rc);
    } else if (error_counts != 0) {
        printf("Error: Rank %d completed with invalid data.\n",world_rank);
    } else {
        printf("Rank %d completed alltoallv successfully: (%f usec).\n",world_rank,1e6*(t1-t0));
    }
    rc |= MPI_Finalize();
    if (!rc) rc = error_counts;
    return rc;
}
```

**Comment by lrbison** [comment]
Well, I should have taken the time to read the standard a little closer before making this PR:

> The "in place" option for intra-communicators is specified by passing MPI_IN_PLACE to
the argument sendbuf at _all_ processes.

Once I fix my test case to abide by this rule, my test case has no hang in main branch.

**Comment by lrbison** [comment]
I can also confirm that main does not hang when I use https://github.com/open-mpi/ompi-tests/blob/master/ibm/collective/alltoallv_somezeros.c

I suppose I can close this then.  @wenduwan are you still seeing hangs?

File: ompi/mca/coll/base/coll_base_alltoallv.c [hunk 1]
```diff
@@ -217,32 +227,72 @@ ompi_coll_base_alltoallv_intra_pairwise(const void *sbuf, const int *scounts, co
     ompi_datatype_type_size(sdtype, &sdtype_size);
     ompi_datatype_type_size(rdtype, &rdtype_size);
 
-    if (0 == sdtype_size || 0 == rdtype_size) {
+    if ((0 == sdtype_size || 0 == rdtype_size) && !SEND_0BYTES_ANYWAYS) {
```
Comment (wenduwan) [inline ompi/mca/coll/base/coll_base_alltoallv.c:230]: While we make this change in alltoallv, I think we should also refactor ialltoallv.

Specifically, my previous 2 changes should be revisited
- https://github.com/open-mpi/ompi/commit/a3a478750b64cbb8368a4fbfbe94eea57b9d996c
- https://github.com/open-mpi/ompi/commit/a7b4d3ebaea7ba2d1825d662de789722416c6ce1

With the solution in this PR, I suggest:
1. Remove `SEND_0BYTES_ANYWAYS`
2. Remove `0 == sdtype_size || 0 == rdtype_size` check - each send and recv should be based on `0 < count && 0 < dtype_size`. I realize my previous change was also wrong in that `sdtype_size` and `rdtype_size` should be evaluated separately.

File: ompi/mca/coll/base/coll_base_alltoallv.c [hunk 2]
```diff
@@ -217,32 +227,72 @@ ompi_coll_base_alltoallv_intra_pairwise(const void *sbuf, const int *scounts, co
...
-                                        prcv, rcounts[recvfrom], rdtype, recvfrom,
-                                        MCA_COLL_BASE_TAG_ALLTOALLV,
-                                        comm, MPI_STATUS_IGNORE, rank);
-        if (MPI_SUCCESS != err) { line = __LINE__; goto err_hndl;  }
+        psnd_left = (char*)sbuf + (ptrdiff_t)sdisps[left] * sext;
+        prcv_left = (char*)rbuf + (ptrdiff_t)rdisps[left] * rext;
+        psnd_right = (char*)sbuf + (ptrdiff_t)sdisps[right] * sext;
+        prcv_right = (char*)rbuf + (ptrdiff_t)rdisps[right] * rext;
+
+        if (0 != rcounts[right]*rdtype_size || SEND_0BYTES_ANYWAYS) {
```
Comment (wenduwan) [inline ompi/mca/coll/base/coll_base_alltoallv.c:261]: `count * dtype_size` is not very efficient. `0 < count && 0 < dtype_size` should be better.

**Review by wenduwan — CHANGES_REQUESTED** [review]

**Comment by bosilca** [comment]
Now that we have the correct understanding of the use of `MPI_IN_PLACE` for alltoallv, what is this PR really solving ?

**Comment by lrbison** [comment]
> Now that we have the correct understanding of the use of `MPI_IN_PLACE` for alltoallv, what is this PR really solving ?

Agreed.  My test case did not reveal any (valid) cases where main-branch hangs.  I will close this PR.  @wenduwan if you are still observing alltoallv hang, in CI or MTT, please ping me.

**Pull request closed without merging**